ci/scripts/run-codebuild/libs/codebuild-runner.js

```diff
@@ -0,0 +1,158 @@
+
+    const commandResponse = await this.codeBuildClient.send(command)
+
+    return commandResponse.builds[0]
+  }
+
+  async _startBuild () {
+    const command = new StartBuildCommand(this.buildCommandInput)
+
+    const startBuildResponse = await this.codeBuildClient.send(command)
+
+    const build = startBuildResponse.build
+
+    this.buildId = build.id
+
+    return build
+  }
+
+  async _waitForBuildToProduceLogs () {
+    console.log(`Waiting for build ${this.buildId} to be in progress`)
+
+    while (true) {
+      const build = await this.getBuildInfo()
+
+      if (build.buildStatus !== 'IN_PROGRESS') {
+        throw new CodeBuildError(`Codebuild finished with status ${build.buildStatus} before producing any logs!`, build)
+      }
+
+      if (typeof build.logs.groupName !== 'undefined') {
+        // Newline gives us a blank line after all the progress .'s
+        console.log(`\nLogs are now availabile at ${build.logs.deepLink} in group ${build.logs.groupName} and stream ${build.logs.streamName}`)
+        return build
+      }
+
+      process.stdout.write('.')
+      await sleep(this.defaultSleepTime)
+    }
+  }
+
+  async _logWhileWaitingForBuildToComplete () {
+    let nextLogToken = null
+
+    while (true) {
+      const buildInfo = await this.getBuildInfo()
+
+      if (buildInfo.buildComplete === true) {
+        // Sleep double the default to give chance for the final logs to get into CloudWatch
+        await sleep(this.defaultSleepTime * 2)
+
+        this._printBuildLogs(buildInfo, nextLogToken)
+
+        if (buildInfo.buildStatus !== 'SUCCEEDED') {
+          throw new CodeBuildError(`Codebuild failed with status ${buildInfo.buildStatus}!`, buildInfo)
+        }
+
+        return buildInfo
+      }
+
+      await sleep(this.defaultSleepTime)
+
+      nextLogToken = await this._printBuildLogs(buildInfo, nextLogToken)
+    }
+  }
+
+  async _printBuildLogs (build, startToken) {
+    const getLogEventsConfig = {
+      logGroupName: build.logs.groupName,
+      logStreamName: build.logs.streamName,
+      startFromHead: true,
+      nextToken: startToken
+    }
+
+    let command = new GetLogEventsCommand(getLogEventsConfig)
+    let response = await this.cloudWatchLogsClient.send(command)
+    let nextToken = null
+
+    while (true) {
+      for (const logEvent of response.events) {
+        // The build logs already have newlines on the end
+        process.stdout.write(logEvent.message)
+      }
+
+      // When the nextForwardToken is the same as the nextToken we passed in
+      // we have reached the end of the stream. This is different than every other
+      // API in the aws-sdk annoyingly where it would return undefined/null as the next
+      // token.
+      if (response.nextForwardToken === nextToken) {
+        return response.nextForwardToken
+      }
+
+      nextToken = response.nextForwardToken
+      command = new GetLogEventsCommand({
+        ...getLogEventsConfig,
```

Comment:
What are these three dots doing?

Reply:
It takes getLogEventsConfig and expands it so every attribte in that becomes an attribute in this, then the nextToken is overridden on the next line. It's like saying "merge everything below on top of getLogEventsConfig"

For a better more formal definition see the MDN docs for javascript spread (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax)